In [24]:
import pandas as pd

excel_file_path = '/content/sample_data/MSCI151 Data_new.xlsx'
df = pd.read_excel(excel_file_path)

print("Heading of the data")
display(df.head())

Heading of the data


id  Profitability   debt     RnD  Firm Age  Firm Size  CEO Experience  \
0   1           0.02  23.76   53.09      3.43      21.22               2   
1   2           3.99  41.85   45.30      3.22      21.83               4   
2   3          41.98  61.84   51.16      3.18      20.54               3   
3   4           8.48  25.61  115.49      3.26      20.59               3   
4   5          -1.94  74.24   13.57      3.26      18.73               3   

   foreign sales  CEO education  CEO Gender  Marketing  Product Sentimet  \
0           9.16              2           0   0.434080          0.443087   
1          41.42              1           0   0.740060          0.860709   
2           5.88              3           1   0.194094          0.306597   
3          11.08              2           0   0.004992          0.400907   
4          24.61              1           0   0.381754          0.797531   

   Politician  PoB  
0    0.141194    0  
1    0.109703    0  
2    0.388709    0  
3    0.221843    0  
4    0.515557    0

In [25]:
print(df.columns)


Index(['id', 'Profitability', 'debt', 'RnD', 'Firm Age', 'Firm Size',
       'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender',
       'Marketing', 'Product Sentimet', 'Politician', 'PoB'],
      dtype='object')


In [26]:
from sklearn.metrics import mean_squared_error, r2_score

# Predict on the test data
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R-squared (R2): {r2:.2f}")

Mean Squared Error (MSE): 194.20
R-squared (R2): -3.22


In [27]:
from sklearn.ensemble import RandomForestRegressor

# Train the Random Forest Regressor model
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

print("Random Forest Regressor model trained successfully.")

Random Forest Regressor model trained successfully.


In [29]:
from sklearn.model_selection import train_test_split

# Prepare the data
features = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Product Sentimet', 'PoB' ]
target = 'Profitability'

X = df[features]
y = df[target]

# Split the data into training and testing sets (80/20 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Data split into training and testing sets.")
print(f"Training set shape: {X_train.shape}, {y_train.shape}")
print(f"Testing set shape: {X_test.shape}, {y_test.shape}")

Data split into training and testing sets.
Training set shape: (97, 11), (97,)
Testing set shape: (25, 11), (25,)


# Task
Perform a predictive modeling task using a Decision Tree Regressor on the dataset from the file "/content/sample_data/MSCI151 Data.xlsx". The target variable is 'Profitability', and the features are 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Product Sentiment', and 'PoB'. Split the data into training and testing sets using an 80/20 ratio.

## Prepare the data

### Subtask:
Select the features (X) and the target variable (y) from the dataframe, using the provided list of features.


**Reasoning**:
Select the features and target variable from the DataFrame based on the provided lists.



In [16]:
features = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Product Sentimet', 'PoB' ]
target = 'Profitability'

X = df[features]
y = df[target]

print("Features (X) and target variable (y) selected.")
display(X.head())
display(y.head())

Features (X) and target variable (y) selected.


debt     RnD  Firm Age  Firm Size  CEO Experience  foreign sales  \
0  23.76   53.09      3.43      21.22               2           9.16   
1  41.85   45.30      3.22      21.83               4          41.42   
2  61.84   51.16      3.18      20.54               3           5.88   
3  25.61  115.49      3.26      20.59               3          11.08   
4  74.24   13.57      3.26      18.73               3          24.61   

   CEO education  CEO Gender  Marketing  Product Sentimet  PoB  
0              2           0   0.434080          0.206488    0  
1              1           0   0.740060          0.208166    0  
2              3           1   0.194094          0.271721    0  
3              2           0   0.004992          0.765643    0  
4              1           0   0.381754          0.679276    0

0     0.02
1     3.99
2    41.98
3     8.48
4    -1.94
Name: Profitability, dtype: float64

## Split the data

### Subtask:
Split the data into training and testing sets using an 80/20 ratio.


**Reasoning**:
The previous attempt to split the data failed due to a KeyError. I need to correct the feature list to match the column names in the dataframe and then split the data into training and testing sets.



In [17]:
from sklearn.model_selection import train_test_split

# Prepare the data - corrected feature list
features = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Product Sentimet', 'PoB' ]
target = 'Profitability'

X = df[features]
y = df[target]

# Split the data into training and testing sets (80/20 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Data split into training and testing sets.")
print(f"Training set shape: {X_train.shape}, {y_train.shape}")
print(f"Testing set shape: {X_test.shape}, {y_test.shape}")

Data split into training and testing sets.
Training set shape: (97, 11), (97,)
Testing set shape: (25, 11), (25,)


## Train the decision tree model

### Subtask:
Train a Decision Tree Regressor model using the training data.


**Reasoning**:
Import the Decision Tree Regressor and train the model using the training data.



In [18]:
from sklearn.tree import DecisionTreeRegressor

# Instantiate a DecisionTreeRegressor object
model = DecisionTreeRegressor(random_state=42)

# Train the Decision Tree Regressor model
model.fit(X_train, y_train)

print("Decision Tree Regressor model trained successfully.")

Decision Tree Regressor model trained successfully.


## Evaluate the model

### Subtask:
Evaluate the Decision Tree model's performance using appropriate metrics on the testing data.


**Reasoning**:
Import the necessary evaluation metrics and calculate the MSE and R2 scores for the model's predictions on the test set.



In [31]:
from sklearn.metrics import mean_squared_error, r2_score

# Predict on the test data
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R-squared (R2): {r2:.2f}")

Mean Squared Error (MSE): 238.11
R-squared (R2): -4.17


In [39]:
# Get feature importances from the trained Decision Tree model
feature_importances = model.feature_importances_

# Get the names of the features
feature_names = X.columns

# Create a pandas Series for better visualization
feature_importance_series = pd.Series(feature_importances, index=feature_names)

# Sort the features by importance in descending order
sorted_feature_importance = feature_importance_series.sort_values(ascending=False)

# Display the feature importances
print("Feature Importances (Decision Tree):")
display(sorted_feature_importance)

Feature Importances (Decision Tree):


Marketing           0.441651
CEO Gender          0.161087
Firm Size           0.112667
foreign sales       0.088469
debt                0.066554
CEO Experience      0.043893
Product Sentimet    0.036848
RnD                 0.033355
Firm Age            0.015330
CEO education       0.000145
dtype: float64

In [38]:
from sklearn.linear_model import Lasso
import numpy as np

# Instantiate a Lasso model
# You might need to tune the alpha parameter based on your data
lasso = Lasso(alpha=0.1, random_state=42)

# Fit the Lasso model to the data
lasso.fit(X, y)

# Get the coefficients of the features
lasso_coefficients = lasso.coef_

# Get the names of the features
feature_names = X.columns

# Create a pandas Series for better visualization
lasso_feature_importance = pd.Series(lasso_coefficients, index=feature_names)

# Display the feature importances (coefficients)
print("Feature Importances (LASSO Regularization):")
display(lasso_feature_importance.sort_values(ascending=False))

Feature Importances (LASSO Regularization):


CEO Gender          9.616555
Product Sentimet    7.499334
Marketing           3.377732
CEO Experience      2.465320
CEO education       1.803266
Firm Size           0.575567
RnD                 0.033668
foreign sales      -0.022050
debt               -0.190809
Firm Age           -2.778382
dtype: float64

## Summary:

### Data Analysis Key Findings

* The data was successfully split into training (97 samples) and testing (25 samples) sets with an 80/20 ratio.
* A Decision Tree Regressor model was trained on the training data.
* The model's performance on the test set resulted in a Mean Squared Error (MSE) of 194.20 and an R-squared (R2) of -3.22.

### Insights or Next Steps

* The negative R-squared value indicates that the current Decision Tree Regressor model performs worse than a simple model that always predicts the mean of the target variable. This suggests significant issues, potentially overfitting or poor feature relevance.
* Explore alternative regression models, feature engineering, or hyperparameter tuning for the Decision Tree Regressor to improve performance.


# Task
Perform a predictive modeling task using a Decision Tree Regressor model. The target variable is 'Profitability', and the features are 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', and 'Product Sentimet'. Split the data into training and testing sets using an 80/20 ratio.

## Prepare the data

### Subtask:
Select the specified features (X) and the target variable (y) from the dataframe.


**Reasoning**:
Select the features and target variable from the DataFrame based on the provided lists.



In [34]:
features = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Product Sentimet' ]
target = 'Profitability'

X = df[features]
y = df[target]

print("Features (X) and target variable (y) selected.")
display(X.head())
display(y.head())

Features (X) and target variable (y) selected.


debt     RnD  Firm Age  Firm Size  CEO Experience  foreign sales  \
0  23.76   53.09      3.43      21.22               2           9.16   
1  41.85   45.30      3.22      21.83               4          41.42   
2  61.84   51.16      3.18      20.54               3           5.88   
3  25.61  115.49      3.26      20.59               3          11.08   
4  74.24   13.57      3.26      18.73               3          24.61   

   CEO education  CEO Gender  Marketing  Product Sentimet  
0              2           0   0.434080          0.443087  
1              1           0   0.740060          0.860709  
2              3           1   0.194094          0.306597  
3              2           0   0.004992          0.400907  
4              1           0   0.381754          0.797531

0     0.02
1     3.99
2    41.98
3     8.48
4    -1.94
Name: Profitability, dtype: float64

## Split the data

### Subtask:
Split the data into training and testing sets using an 80/20 ratio.


**Reasoning**:
Import the train_test_split function and split the data into training and testing sets.



In [35]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets (80/20 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Data split into training and testing sets.")
print(f"Training set shape: {X_train.shape}, {y_train.shape}")
print(f"Testing set shape: {X_test.shape}, {y_test.shape}")

Data split into training and testing sets.
Training set shape: (97, 10), (97,)
Testing set shape: (25, 10), (25,)


## Train the decision tree model

### Subtask:
Train a Decision Tree Regressor model using the training data.


**Reasoning**:
Import the Decision Tree Regressor and train the model using the training data.



In [36]:
from sklearn.tree import DecisionTreeRegressor

# Instantiate a DecisionTreeRegressor object
model = DecisionTreeRegressor(random_state=42)

# Train the Decision Tree Regressor model
model.fit(X_train, y_train)

print("Decision Tree Regressor model trained successfully.")

Decision Tree Regressor model trained successfully.


## Evaluate the model

### Subtask:
Evaluate the Decision Tree model's performance using appropriate metrics on the testing data.


**Reasoning**:
Import the necessary evaluation metrics and calculate the MSE and R2 scores for the model's predictions on the test set.



In [37]:
from sklearn.metrics import mean_squared_error, r2_score

# Predict on the test data
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R-squared (R2): {r2:.2f}")

Mean Squared Error (MSE): 230.63
R-squared (R2): -4.01


## Summary:

### Data Analysis Key Findings

*   The data was successfully split into training (97 samples) and testing (25 samples) sets using an 80/20 ratio.
*   A Decision Tree Regressor model was trained on the training data.
*   The model's performance on the testing data was evaluated using Mean Squared Error (MSE) and R-squared ($\text{R}^2$).
*   The Mean Squared Error (MSE) was 230.63.
*   The R-squared ($\text{R}^2$) score was -4.01, indicating that the model performs worse than predicting the mean of the target variable.

### Insights or Next Steps

*   The current Decision Tree Regressor model performs poorly, as indicated by the high MSE and significantly negative $\text{R}^2$.
*   Further steps should involve exploring other regression models, tuning hyperparameters of the Decision Tree, or investigating feature engineering to improve model performance.


# Task
Perform an OLS regression with intercept and robust standard errors using the data from "/content/sample_data/MSCI151 Data.xlsx". The independent variables (X) are 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'marketing', 'Product Sentiment', and 'PoB', and the dependent variable (Y) is 'Profitability'.

## Prepare the data

### Subtask:
Define the features (X) and the target variable (y) from the dataframe.


**Reasoning**:
Define the features and target variable from the dataframe.



In [40]:
features = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Product Sentimet', 'PoB' ]
target = 'Profitability'

X = df[features]
y = df[target]

print("Features (X) and target variable (y) selected.")
display(X.head())
display(y.head())

Features (X) and target variable (y) selected.


debt     RnD  Firm Age  Firm Size  CEO Experience  foreign sales  \
0  23.76   53.09      3.43      21.22               2           9.16   
1  41.85   45.30      3.22      21.83               4          41.42   
2  61.84   51.16      3.18      20.54               3           5.88   
3  25.61  115.49      3.26      20.59               3          11.08   
4  74.24   13.57      3.26      18.73               3          24.61   

   CEO education  CEO Gender  Marketing  Product Sentimet  PoB  
0              2           0   0.434080          0.443087    0  
1              1           0   0.740060          0.860709    0  
2              3           1   0.194094          0.306597    0  
3              2           0   0.004992          0.400907    0  
4              1           0   0.381754          0.797531    0

0     0.02
1     3.99
2    41.98
3     8.48
4    -1.94
Name: Profitability, dtype: float64

## Add a constant (intercept)

### Subtask:
Add a constant term to the features DataFrame for the intercept in the OLS model.


**Reasoning**:
Add a constant term to the features DataFrame for the intercept in the OLS model.



In [41]:
import statsmodels.api as sm

# Add a constant term to the features DataFrame for the intercept
X = sm.add_constant(X)

print("Constant term added to the features DataFrame.")
display(X.head())

Constant term added to the features DataFrame.


const   debt     RnD  Firm Age  Firm Size  CEO Experience  foreign sales  \
0    1.0  23.76   53.09      3.43      21.22               2           9.16   
1    1.0  41.85   45.30      3.22      21.83               4          41.42   
2    1.0  61.84   51.16      3.18      20.54               3           5.88   
3    1.0  25.61  115.49      3.26      20.59               3          11.08   
4    1.0  74.24   13.57      3.26      18.73               3          24.61   

   CEO education  CEO Gender  Marketing  Product Sentimet  PoB  
0              2           0   0.434080          0.443087    0  
1              1           0   0.740060          0.860709    0  
2              3           1   0.194094          0.306597    0  
3              2           0   0.004992          0.400907    0  
4              1           0   0.381754          0.797531    0

## Fit the ols model with robust standard errors

### Subtask:
Use `statsmodels.api` to fit the OLS model with the specified features and target, and calculate robust standard errors.


**Reasoning**:
Use statsmodels.api to fit the OLS model with the specified features and target, and calculate robust standard errors.



In [42]:
# Fit the OLS model
model = sm.OLS(y, X)
results = model.fit()

# Calculate robust standard errors (HC1)
robust_se = results.get_robust_cov_params(cov_type='HC1')

print("OLS model fitted and robust standard errors calculated.")
display(results.summary())
display(robust_se)

AttributeError: 'OLSResults' object has no attribute 'get_robust_cov_params'

**Reasoning**:
The previous attempt to get robust standard errors failed because the method `get_robust_cov_params` was not found. I need to use the correct method to calculate robust standard errors, which is by specifying the `cov_type` argument directly in the `fit()` method.



In [43]:
# Fit the OLS model with robust standard errors (HC1)
results = model.fit(cov_type='HC1')

print("OLS model fitted with robust standard errors.")
display(results.summary())

OLS model fitted with robust standard errors.


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          Profitability   R-squared:                       0.318
Model:                            OLS   Adj. R-squared:                  0.249
Method:                 Least Squares   F-statistic:                     4.216
Date:                Wed, 08 Oct 2025   Prob (F-statistic):           3.28e-05
Time:                        05:30:09   Log-Likelihood:                -462.19
No. Observations:                 122   AIC:                             948.4
Df Residuals:                     110   BIC:                             982.0
Df Model:                          11                                         
Covariance Type:                  HC1                                         
====================================================================================
                       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------
const              -11.5778     15.744     -0.735      0.462     -42.435      19.279
debt                -0.1848      0.081     -2.291      0.022      -0.343      -0.027
RnD                  0.0327      0.020      1.678      0.093      -0.005       0.071
Firm Age            -3.6668      2.711     -1.353      0.176      -8.980       1.646
Firm Size            0.6688      0.714      0.937      0.349      -0.730       2.068
CEO Experience       3.1706      2.335      1.358      0.175      -1.407       7.748
foreign sales       -0.0269      0.029     -0.923      0.356      -0.084       0.030
CEO education        2.1024      2.444      0.860      0.390      -2.688       6.893
CEO Gender          10.0401      2.742      3.662      0.000       4.666      15.414
Marketing            4.5876      3.987      1.151      0.250      -3.227      12.402
Product Sentimet     8.9679      5.216      1.719      0.086      -1.256      19.192
PoB                  0.6684      2.109      0.317      0.751      -3.464       4.801
==============================================================================
Omnibus:                      101.200   Durbin-Watson:                   2.018
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1470.078
Skew:                          -2.606   Prob(JB):                         0.00
Kurtosis:                      19.187   Cond. No.                     1.16e+03
==============================================================================

Notes:
[1] Standard Errors are heteroscedasticity robust (HC1)
[2] The condition number is large, 1.16e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

## Display the regression results

### Subtask:
Print the summary of the regression results, including coefficients, standard errors, p-values, and R-squared.


**Reasoning**:
Display the summary of the results object to show the regression output with robust standard errors.



In [49]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:          Profitability   R-squared:                       0.318
Model:                            OLS   Adj. R-squared:                  0.249
Method:                 Least Squares   F-statistic:                     4.216
Date:                Wed, 08 Oct 2025   Prob (F-statistic):           3.28e-05
Time:                        05:40:39   Log-Likelihood:                -462.19
No. Observations:                 122   AIC:                             948.4
Df Residuals:                     110   BIC:                             982.0
Df Model:                          11                                         
Covariance Type:                  HC1                                         
                       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------
const              -11.5778     15.744  

## Summary:

### Data Analysis Key Findings

*   The R-squared value for the model is 0.318, indicating that approximately 31.8% of the variance in 'Profitability' is explained by the independent variables.
*   'debt' has a statistically significant negative relationship with 'Profitability' (p-value = 0.022).
*   'CEO Gender' has a statistically significant positive relationship with 'Profitability' (p-value = 0.000).
*   'RnD' and 'Product Sentiment' show a marginally significant positive relationship with 'Profitability' (p-values = 0.093 and 0.086, respectively).
*   The standard errors used in the model are heteroscedasticity robust (HC1).
*   The model summary indicates a large condition number (1.16e+03), suggesting potential multicollinearity among the independent variables.

### Insights or Next Steps

*   Investigate the potential multicollinearity among the independent variables, possibly by examining correlation matrices or using variance inflation factors (VIFs), and consider strategies to address it if necessary (e.g., variable removal, principal component analysis).
*   Explore the nature of the significant relationships further, particularly for 'debt' and 'CEO Gender', to understand the practical implications of these findings for firm profitability.


# Task
Re-run a predictive model under the decision tree and its feature importance using the data from "/content/sample_data/MSCI151 Data.xlsx", where the Y is 'Profitability', and the Xs are: 'debt', 'RnD', 'CEO Gender', 'Product Sentiment'.

## Prepare the data

### Subtask:
Select the specified features (X) and the target variable (y) from the dataframe.


**Reasoning**:
Select the features and target variable from the DataFrame based on the provided lists.



In [50]:
features = ['debt', 'RnD', 'CEO Gender', 'Product Sentimet' ]
target = 'Profitability'

X = df[features]
y = df[target]

print("Features (X) and target variable (y) selected.")
display(X.head())
display(y.head())

Features (X) and target variable (y) selected.


debt     RnD  CEO Gender  Product Sentimet
0  23.76   53.09           0          0.443087
1  41.85   45.30           0          0.860709
2  61.84   51.16           1          0.306597
3  25.61  115.49           0          0.400907
4  74.24   13.57           0          0.797531

0     0.02
1     3.99
2    41.98
3     8.48
4    -1.94
Name: Profitability, dtype: float64

## Split the data

### Subtask:
Split the data into training and testing sets using an 80/20 ratio.


**Reasoning**:
Import the train_test_split function and split the data into training and testing sets.



In [51]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets (80/20 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Data split into training and testing sets.")
print(f"Training set shape: {X_train.shape}, {y_train.shape}")
print(f"Testing set shape: {X_test.shape}, {y_test.shape}")

Data split into training and testing sets.
Training set shape: (97, 4), (97,)
Testing set shape: (25, 4), (25,)


## Train the decision tree model

### Subtask:
Train a Decision Tree Regressor model using the training data.


**Reasoning**:
Import the Decision Tree Regressor and train the model using the training data.



In [52]:
from sklearn.tree import DecisionTreeRegressor

# Instantiate a DecisionTreeRegressor object
model = DecisionTreeRegressor(random_state=42)

# Train the Decision Tree Regressor model
model.fit(X_train, y_train)

print("Decision Tree Regressor model trained successfully.")

Decision Tree Regressor model trained successfully.


## Calculate and display feature importances

### Subtask:
Get the feature importances from the trained model and display them.


**Reasoning**:
Get the feature importances from the trained model and display them.



In [53]:
# Get feature importances from the trained Decision Tree model
feature_importances = model.feature_importances_

# Get the names of the features
feature_names = X_train.columns

# Create a pandas Series for better visualization
feature_importance_series = pd.Series(feature_importances, index=feature_names)

# Sort the features by importance in descending order
sorted_feature_importance = feature_importance_series.sort_values(ascending=False)

# Display the feature importances
print("Feature Importances (Decision Tree):")
display(sorted_feature_importance)

Feature Importances (Decision Tree):


Product Sentimet    0.417552
debt                0.325753
CEO Gender          0.145482
RnD                 0.111213
dtype: float64

## Evaluate the model (optional but recommended)

### Subtask:
Evaluate the model's performance using appropriate metrics on the testing data to see if the reduced feature set improves or worsens performance.


**Reasoning**:
Import the necessary evaluation metrics and calculate the MSE and R2 scores for the model's predictions on the test set.



In [54]:
from sklearn.metrics import mean_squared_error, r2_score

# Predict on the test data
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R-squared (R2): {r2:.2f}")

Mean Squared Error (MSE): 130.64
R-squared (R2): -1.84


## Summary:

### Data Analysis Key Findings

*   The Decision Tree Regressor model was trained using 'debt', 'RnD', 'CEO Gender', and 'Product Sentiment' as features and 'Profitability' as the target.
*   The feature importances from the trained model indicate that 'Product Sentiment' is the most important feature for predicting 'Profitability', followed by 'debt', 'CEO Gender', and 'RnD'.
*   The model evaluation on the test set resulted in a Mean Squared Error (MSE) of 130.64 and an R-squared (R2) score of -1.84.

### Insights or Next Steps

*   The negative R2 score indicates that the model's performance is very poor and worse than simply predicting the mean of the target variable. This suggests that the selected features may not be sufficiently predictive of profitability, or that the Decision Tree model is not the most suitable algorithm for this dataset with these features.
*   Further steps should involve exploring other regression models, performing feature engineering, or considering a different set of features to improve the model's performance.


# Task
Perform Principal Component Analysis (PCA) on the dataset loaded from "/content/sample_data/MSCI151 Data.xlsx". The independent variables (X) for the PCA are 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'marketing', 'Product Sentiment', and 'PoB'. The dependent variable (Y) is 'Profitability'. Provide the loading factors for each variable.

## Prepare the data

### Subtask:
Select the specified features (X) from the dataframe for PCA.


**Reasoning**:
Define the features and select them from the dataframe.



In [55]:
features = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Product Sentimet', 'PoB' ]
X = df[features]

print("Features (X) selected for PCA.")
display(X.head())

Features (X) selected for PCA.


debt     RnD  Firm Age  Firm Size  CEO Experience  foreign sales  \
0  23.76   53.09      3.43      21.22               2           9.16   
1  41.85   45.30      3.22      21.83               4          41.42   
2  61.84   51.16      3.18      20.54               3           5.88   
3  25.61  115.49      3.26      20.59               3          11.08   
4  74.24   13.57      3.26      18.73               3          24.61   

   CEO education  CEO Gender  Marketing  Product Sentimet  PoB  
0              2           0   0.434080          0.443087    0  
1              1           0   0.740060          0.860709    0  
2              3           1   0.194094          0.306597    0  
3              2           0   0.004992          0.400907    0  
4              1           0   0.381754          0.797531    0

## Standardize the data

### Subtask:
Standardize the feature data since PCA is affected by scale.


**Reasoning**:
Standardize the feature data for PCA.



In [56]:
from sklearn.preprocessing import StandardScaler

# Instantiate a StandardScaler object
scaler = StandardScaler()

# Fit the scaler to the features DataFrame X
scaler.fit(X)

# Transform the features DataFrame X and store the standardized data
X_scaled = scaler.transform(X)

# Convert the scaled data back to a DataFrame for easier viewing (optional)
X_scaled = pd.DataFrame(X_scaled, columns=features)


print("Features (X) standardized.")
display(X_scaled.head())

Features (X) standardized.


debt       RnD  Firm Age  Firm Size  CEO Experience  foreign sales  \
0 -1.157075  0.572487  0.219984   1.397541       -2.106561      -1.399088   
1 -0.067926  0.414239 -0.218273   1.767876        1.366418      -0.312128   
2  1.135618  0.533280 -0.301750   0.984708       -0.370072      -1.509603   
3 -1.045692  1.840095 -0.134796   1.015063       -0.370072      -1.334396   
4  1.882188 -0.230332 -0.134796  -0.114156       -0.370072      -0.878520   

   CEO education  CEO Gender  Marketing  Product Sentimet       PoB  
0       1.099711   -0.469042  -0.156763         -0.198992 -0.777987  
1      -0.713326   -0.469042   0.867755          1.269842 -0.777987  
2       2.912748    2.132007  -0.960309         -0.679043 -0.777987  
3       1.099711   -0.469042  -1.593482         -0.347344 -0.777987  
4      -0.713326   -0.469042  -0.331966          1.047637 -0.777987

## Perform pca

### Subtask:
Apply PCA to the standardized data stored in `X_scaled`.


**Reasoning**:
Import PCA, instantiate the model, fit it to the scaled data, and transform the scaled data.



In [57]:
from sklearn.decomposition import PCA

# Instantiate a PCA object, keeping all components
pca = PCA(random_state=42)

# Fit PCA to the standardized data
pca.fit(X_scaled)

# Transform the standardized data
X_pca = pca.transform(X_scaled)

print("PCA applied to the standardized data.")
print(f"Shape of transformed data: {X_pca.shape}")

PCA applied to the standardized data.
Shape of transformed data: (122, 11)


## Get loading factors

### Subtask:
Calculate and display the loading factors for each variable on the principal components.


**Reasoning**:
Access the principal components from the fitted PCA object and create a DataFrame to display the loading factors with appropriate labels.



In [58]:
import pandas as pd

# Access the principal components (eigenvectors)
pca_components = pca.components_

# Create a DataFrame for better visualization of loading factors
# Rows are principal components, columns are original features
loading_factors = pd.DataFrame(pca_components.T, columns=[f'Principal Component {i+1}' for i in range(pca_components.shape[0])], index=features)

# Display the loading factors
print("Loading Factors for each variable on the Principal Components:")
display(loading_factors)

Loading Factors for each variable on the Principal Components:


Principal Component 1  Principal Component 2  \
debt                          -0.157546               0.116967   
RnD                            0.254795               0.113019   
Firm Age                       0.399879               0.154343   
Firm Size                      0.380402               0.376875   
CEO Experience                 0.230098               0.476379   
foreign sales                 -0.154713              -0.058783   
CEO education                  0.510161              -0.372304   
CEO Gender                     0.510381              -0.366366   
Marketing                     -0.042617              -0.439170   
Product Sentimet               0.001253              -0.130524   
PoB                           -0.078725              -0.307500   

                  Principal Component 3  Principal Component 4  \
debt                          -0.459683              -0.233422   
RnD                            0.064064              -0.456538   
Firm Age                       0.179518              -0.134240   
Firm Size                     -0.168892               0.296677   
CEO Experience                 0.390202               0.128787   
foreign sales                  0.249376               0.614884   
CEO education                 -0.253680              -0.096682   
CEO Gender                     0.055760               0.234638   
Marketing                      0.129308              -0.009166   
Product Sentimet              -0.400445               0.370922   
PoB                            0.514636              -0.185235   

                  Principal Component 5  Principal Component 6  \
debt                           0.123664               0.582846   
RnD                           -0.468266               0.207323   
Firm Age                       0.582831              -0.081337   
Firm Size                     -0.005076              -0.179773   
CEO Experience                 0.131834               0.313880   
foreign sales                 -0.210948               0.467853   
CEO education                 -0.057476               0.164820   
CEO Gender                    -0.215578               0.102417   
Marketing                      0.542108               0.315154   
Product Sentimet               0.110728              -0.273208   
PoB                           -0.088986              -0.222141   

                  Principal Component 7  Principal Component 8  \
debt                          -0.317392               0.312292   
RnD                            0.487518               0.226393   
Firm Age                      -0.136283               0.370980   
Firm Size                     -0.248183              -0.124327   
CEO Experience                 0.255772               0.022981   
foreign sales                 -0.131631               0.263146   
CEO education                 -0.208025               0.020072   
CEO Gender                    -0.018201              -0.116366   
Marketing                      0.343648              -0.262767   
Product Sentimet               0.486611               0.570941   
PoB                           -0.316520               0.468400   

                  Principal Component 9  Principal Component 10  \
debt                           0.016349                0.315173   
RnD                            0.322779               -0.194041   
Firm Age                      -0.112784               -0.435086   
Firm Size                      0.676032                0.169052   
CEO Experience                -0.260401                0.437154   
foreign sales                  0.137577               -0.403586   
CEO education                 -0.058208               -0.124175   
CEO Gender                    -0.279715                0.282369   
Marketing                      0.445530                0.095245   
Product Sentimet              -0.017918                0.173264   
PoB                            0.242122                0.396115   

                  Principal Component 11 

## Summary:

### Data Analysis Key Findings

*   The standardized data for PCA has a shape of (122, 11), indicating 122 samples and 11 features.
*   The loading factors were successfully calculated and displayed, showing the contribution of each original variable to each principal component.

### Insights or Next Steps

*   Analyze the loading factors to understand which original variables contribute most to each principal component.
*   Examine the explained variance ratio of the principal components to determine how many components are needed to capture a sufficient amount of the data's variance.
